### PR TITLE
ArrayItemRemoval should not mutate lists

### LINKED ISSUE #1357: In Assert.php line 2042: Expected a value other than null.
| Question    | Answer
| ------------| ---------------
| Infection version | 0.18.0
| Test Framework version | PHPUnit
| PHP version | 7.4.10
| Platform    | windows-github/actions
| Github Repo | https://github.com/yiisoft/active-record/runs/1270725926?check_suite_focus=true

```php
Processing source code files: 0
........U.........T..........T...............UU..U   (   50)
UU............................U...........U.......   (  100)
.......................U...U...UUUUUUUUUUUUUUUUUUU   (  150)
UUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUUU..   (  200)
...............U.....................UU...........   (  250)
....M................UU...........................   (  300)
......M.............................U.......U....U   (  350)
...............U.........U......U.........U.....UU   (  400)

In Assert.php line 2042:
                                     
  Expected a value other than null.  
                                     

run [--test-framework TEST-FRAMEWORK] [--test-framework-options TEST-FRAMEWORK-OPTIONS] [-j|--threads THREADS] [--only-covered] [-s|--show-mutations] [--no-progress] [--force-progress] [-c|--configuration CONFIGURATION] [--coverage COVERAGE] [--mutators MUTATORS] [--filter FILTER] [--formatter FORMATTER] [--min-msi MIN-MSI] [--min-covered-msi MIN-COVERED-MSI] [--log-verbosity LOG-VERBOSITY] [--initial-tests-php-options INITIAL-TESTS-PHP-OPTIONS] [--skip-initial-tests] [--ignore-msi-with-no-mutations] [--debug] [--dry-run]

UUUSS.SSS
Error: Process completed with exit code 1.
```

### PR DESCRIPTION
**Pull request by sanmai**
This PR:

- [x] Ensures ArrayItemRemoval does not mutate lists (arrays to the left of an assignment)
- [x] Covered by tests
- [x] Fixes #1357

### CODE REVIEW

File: src/Mutator/Removal/ArrayItemRemoval.php [hunk 1]
```diff
@@ -119,7 +120,22 @@ public function mutate(Node $arrayNode): iterable
 
     public function canMutate(Node $node): bool
     {
-        return $node instanceof Node\Expr\Array_ && count($node->items);
+        if (!$node instanceof Node\Expr\Array_) {
+            return false;
+        }
+
+        if (!count($node->items)) {
```
Comment (maks-rafalko) [inline src/Mutator/Removal/ArrayItemRemoval.php:127]: what do you think about `count() === 0` or `$node->items === []` as we have in other places?

**Review by maks-rafalko — APPROVED** [review]
Reminder for myself: port it to 0.18

File: src/Mutator/Removal/ArrayItemRemoval.php (hunk 1 above)
Comment (sanmai) [inline src/Mutator/Removal/ArrayItemRemoval.php:127]:
```suggestion
        if (count($node->items) === 0) {
```

Comment (sanmai) [reply src/Mutator/Removal/ArrayItemRemoval.php]: Good point.

File: src/Mutator/Removal/ArrayItemRemoval.php [hunk 2]
```diff
@@ -119,7 +120,22 @@ public function mutate(Node $arrayNode): iterable
 
     public function canMutate(Node $node): bool
     {
-        return $node instanceof Node\Expr\Array_ && count($node->items);
+        if (!$node instanceof Node\Expr\Array_) {
+            return false;
+        }
+
+        if (count($node->items) === 0) {
```
Comment (sanmai) [inline src/Mutator/Removal/ArrayItemRemoval.php:127]:
```suggestion
        if ($node->items === []) {
```

**Comment by sanmai** [comment]
@maks-rafalko Can we disable this "branch must be up to date" requirement, what do you think? I had to restart the build two times already...

**Comment by maks-rafalko** [comment]
> @maks-rafalko Can we disable this "branch must be up to date" requirement, what do you think? I had to restart the build two times already...

ok, disabled. This will potentially increase the number of failed builds in master, but let's see

**Comment by maks-rafalko** [comment]
Thank you @sanmai

**Comment by maks-rafalko** [comment]
Ported to 0.18, released as 0.18.1

**Pull request merged**